### PR TITLE
detect/detection_filter: extend unique_on with src_ip|dst_ip - v8

### PR DESCRIPTION
**Pull request by oferda4**
Add optional unique_on {src_ip|dst_ip} to detection_filter for distinct IP address counting within the seconds window. Features:

Runtime uses a hash table per threshold entry for tracking unique IP addresses (both IPv4 and IPv6).
Follows detection_filter semantics: alerting starts after the threshold (>count), not at it.
On window expiry, the window is reset and the current packet's IP is recorded as the first distinct of the new window. Validation:
unique_on src_ip/dst_ip works with any IP protocol (unlike port-based unique_on which requires tcp/udp/sctp). Memory management:
Hash table memory is bounded by detect.thresholds.memcap.
Reuses existing counters: bitmap_memuse and bitmap_alloc_fail. Refactoring:
Added ThresholdDistinctAdd helper to consolidate port/IP tracking logic and reduce code duplication. Tests:
C unit tests for parsing unique_on src_ip and dst_ip options.

Changes:
v2:
- Update documentation
- Increase unit tests coverage
- Document use of hash table

v3:
- Cleanup unit tests

v4:
- Added validation to reject duplicate unique_on options
- Fixed memory accounting in ThresholdDistinctReset: counters are now properly decremented if hash table re-init fails
- Hash table is now pre-sized to count buckets instead of a fixed 64

v5:
- Fixed use-after-free: HashTableAdd stores pointers without copying, so ThresholdDistinctAddIP now allocates owned copies of IP address bytes. A Free callback (ThresholdDistinctIPFree) is registered with the hash table to automatically free copies on reset/free.
- Handle HashTableAdd allocation failure by freeing the IP copy
- Simplified window expiry: te->current_count = 1 is now unconditional
- Added documentation on two-phase memory allocation model (init-time bucket array vs. runtime per-entry copies)

v6:
- Replace HashTable with pre-allocated flat buffer (Address *distinct_ip_buf) for IP deduplication — eliminates all per-packet SCMalloc calls
- Add 4 new unit tests: IPv4 basic + buffer-full boundary, duplicate suppression, window reset, dst_ip with track by_src
- Restrict port-protocol validation to unique_on src_port/dst_port only (not src_ip/dst_ip)

v7:
- Simplified `else if` to `else` in ThresholdDistinctInit (only remaining case after port branch)
- Clarified scan_limit comment: current_count can exceed max_count (counts all distinct IPs seen, but only max_count are stored in the buffer)
- Added DF_UNIQUE_IP_MAX_COUNT (10000) cap for IP-based unique_on to bound per-entry buffer size

v8:
- ThresholdDistinctAdd now returns void; the classic-counting fallback (current_count++) is handled inside the helper, removing the bool return and simplifying the call sites
- Moved DF_UNIQUE_IP_MAX_COUNT define into detect-detection-filter.c (its only user) instead of the shared header
- Lowered DF_UNIQUE_IP_MAX_COUNT cap from 10000 to 1024

Previous PR: https://github.com/OISF/suricata/pull/15194

Ticket: https://redmine.openinfosecfoundation.org/issues/8250
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2924

### CODE REVIEW

File: doc/userguide/rules/thresholding.rst [hunk 1]
```diff
@@ -179,18 +179,24 @@ internal counter and alert each time the threshold has been reached.
...
 .. note::
 
-   ``unique_on`` requires a transport protocol with ports. Use it only with
+   ``unique_on src_port|dst_port`` requires a transport protocol with ports. Use it only with
    TCP, UDP or SCTP rules. For non-ported protocols (or ``ip`` rules), Suricata
-   rejects ``unique_on`` during rule parsing.
+   rejects port-based ``unique_on`` during rule parsing.
+
+   ``unique_on src_ip|dst_ip`` works with any IP protocol. The ``count`` value
+   is capped at 10000 for IP-based ``unique_on`` because each tracked host
```
Comment (jlucovsky) [inline doc/userguide/rules/thresholding.rst:198]: `#define DF_UNIQUE_IP_MAX_COUNT 1024`

This is the cap, not `10000`

File: src/detect-detection-filter.c [hunk 2]
```diff
@@ -173,10 +176,21 @@ static DetectThresholdData *DetectDetectionFilterParse(const char *rawstr)
             count_pos = i + 1;
         if (strncasecmp(args[i], "seconds", strlen("seconds")) == 0)
             seconds_pos = i + 1;
-        if (strcasecmp(args[i], "src_port") == 0)
-            df->unique_on = DF_UNIQUE_SRC_PORT;
-        if (strcasecmp(args[i], "dst_port") == 0)
-            df->unique_on = DF_UNIQUE_DST_PORT;
+        if (strcasecmp(args[i], "src_port") == 0 || strcasecmp(args[i], "dst_port") == 0 ||
```
Comment (jlucovsky) [inline src/detect-detection-filter.c:179]: Will the PCRE allow a second `unique_on` to appear?

File: src/detect-detection-filter.c [hunk 3]
```diff
@@ -1002,6 +1079,375 @@ static int DetectDetectionFilterDistinctBitmapExpiry(void)
...
+    UTHFreePackets(&p1, 1);
+    UTHFreePackets(&p2, 1);
+    UTHFreePackets(&p3, 1);
+    ThresholdDestroy();
+    StatsThreadCleanup(&th_v.stats);
+    PASS;
+}
+
+/**
+ * \test Verify IP hash memory is tracked and pre-sized to count
```
Comment (jlucovsky) [inline src/detect-detection-filter.c:1137]: nit: is hash the correct word?

**Review by jlucovsky — COMMENTED** [review]
Please ensure that the copyright year is updated for modified source modules.

File: src/detect-detection-filter.c (hunk 2 above)
Comment (oferda4) [reply src/detect-detection-filter.c:179]: no: the $-anchored regex rejects a duplicate outright (there's also a defensive guard at line 181-184). No existing parse-rejection test though. I'll add.

File: src/detect-detection-filter.c (hunk 3 above)
Comment (oferda4) [reply src/detect-detection-filter.c:1137]: No you are right. stale wording from old iteration.

**Review by catenacyber — CHANGES_REQUESTED** [review]
Some changes were requested inline
